In [181]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

Neste caderno notebook, o objetivo é realizar uma análise de dados sobre o preço da gasolina.

*Os datasets estão separados por quetão didática, mas segue abaixo o link:https://www.kaggle.com/datasets/matheusfreitag/gas-prices-in-brazil

In [182]:
df1 = pd.read_csv("Datasets/gasolina_2000+.csv", index_col=0)
df2 = pd.read_csv("Datasets/gasolina_2010+.csv", index_col=0, low_memory=False)

df = pd.concat([df1, df2])

In [183]:
df.head(3)

DATA INICIAL  DATA FINAL        REGIÃO            ESTADO           PRODUTO  \
0   2004-05-09  2004-05-15  CENTRO OESTE  DISTRITO FEDERAL  ETANOL HIDRATADO   
1   2004-05-09  2004-05-15  CENTRO OESTE             GOIAS  ETANOL HIDRATADO   
2   2004-05-09  2004-05-15  CENTRO OESTE       MATO GROSSO  ETANOL HIDRATADO   

   NÚMERO DE POSTOS PESQUISADOS UNIDADE DE MEDIDA  PREÇO MÉDIO REVENDA  \
0                           127              R$/l                1.288   
1                           387              R$/l                1.162   
2                           192              R$/l                1.389   

   DESVIO PADRÃO REVENDA  PREÇO MÍNIMO REVENDA  PREÇO MÁXIMO REVENDA  \
0                  0.016                  1.19                 1.350   
1                  0.114                  0.89                 1.449   
2                  0.097                  1.18                 1.760   

  MARGEM MÉDIA REVENDA  COEF DE VARIAÇÃO REVENDA PREÇO MÉDIO DISTRIBUIÇÃO  \
0                0.463                     0.012                    0.825   
1                0.399                     0.098                    0.763   
2                0.419                     0.070                     0.97   

  DESVIO PADRÃO DISTRIBUIÇÃO PREÇO MÍNIMO DISTRIBUIÇÃO  \
0                       0.11                    0.4201   
1                      0.088                    0.5013   
2                      0.095                    0.5614   

  PREÇO MÁXIMO DISTRIBUIÇÃO COEF DE VARIAÇÃO DISTRIBUIÇÃO  
0                    0.9666                         0.133  
1                      1.05                         0.115  
2                     1.161                         0.098

__Conhecendo o conjunto de dados__

In [184]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 120823 entries, 0 to 120822
Data columns (total 18 columns):
 #   Column                         Non-Null Count   Dtype  
---  ------                         --------------   -----  
 0   DATA INICIAL                   120823 non-null  object 
 1   DATA FINAL                     120823 non-null  object 
 2   REGIÃO                         120823 non-null  object 
 3   ESTADO                         120823 non-null  object 
 4   PRODUTO                        120823 non-null  object 
 5   NÚMERO DE POSTOS PESQUISADOS   120823 non-null  int64  
 6   UNIDADE DE MEDIDA              120823 non-null  object 
 7   PREÇO MÉDIO REVENDA            120823 non-null  float64
 8   DESVIO PADRÃO REVENDA          120823 non-null  float64
 9   PREÇO MÍNIMO REVENDA           120823 non-null  float64
 10  PREÇO MÁXIMO REVENDA           120823 non-null  float64
 11  MARGEM MÉDIA REVENDA           120823 non-null  object 
 12  COEF DE VARIAÇÃO REVENDA      

__Conversão de tipos__

_Alteração dos tipos nas colunas

* 0   DATA INICIAL                   120823 non-null  object  Datetime
* 1   DATA FINAL                     120823 non-null  object  Datetime
* 2   REGIÃO                         120823 non-null  object  Correto
* 3   ESTADO                         120823 non-null  object  Correto
* 4   PRODUTO                        120823 non-null  object  Correto
* 5   NÚMERO DE POSTOS PESQUISADOS   120823 non-null  int64   Correto
* 6   UNIDADE DE MEDIDA              120823 non-null  object  Inteiro
* 7   PREÇO MÉDIO REVENDA            120823 non-null  float64 Correto
* 8   DESVIO PADRÃO REVENDA          120823 non-null  float64 Correto
* 9   PREÇO MÍNIMO REVENDA           120823 non-null  float64 Correto
* 10  PREÇO MÁXIMO REVENDA           120823 non-null  float64 Correto
* 11  MARGEM MÉDIA REVENDA           120823 non-null  object  Float
* 12  COEF DE VARIAÇÃO REVENDA       120823 non-null  float64 Correto
* 13  PREÇO MÉDIO DISTRIBUIÇÃO       120823 non-null  object  Float
* 14  DESVIO PADRÃO DISTRIBUIÇÃO     120823 non-null  object  Float
* 15  PREÇO MÍNIMO DISTRIBUIÇÃO      120823 non-null  object  Float
* 16  PREÇO MÁXIMO DISTRIBUIÇÃO      120823 non-null  object  Float
* 17  COEF DE VARIAÇÃO DISTRIBUIÇÃO  120823 non-null  object  Float
* dtypes: float64(5), int64(1), object(12)

In [185]:
df.replace('-',0, inplace=True)
df.head(3)

DATA INICIAL  DATA FINAL        REGIÃO            ESTADO           PRODUTO  \
0   2004-05-09  2004-05-15  CENTRO OESTE  DISTRITO FEDERAL  ETANOL HIDRATADO   
1   2004-05-09  2004-05-15  CENTRO OESTE             GOIAS  ETANOL HIDRATADO   
2   2004-05-09  2004-05-15  CENTRO OESTE       MATO GROSSO  ETANOL HIDRATADO   

   NÚMERO DE POSTOS PESQUISADOS UNIDADE DE MEDIDA  PREÇO MÉDIO REVENDA  \
0                           127              R$/l                1.288   
1                           387              R$/l                1.162   
2                           192              R$/l                1.389   

   DESVIO PADRÃO REVENDA  PREÇO MÍNIMO REVENDA  PREÇO MÁXIMO REVENDA  \
0                  0.016                  1.19                 1.350   
1                  0.114                  0.89                 1.449   
2                  0.097                  1.18                 1.760   

  MARGEM MÉDIA REVENDA  COEF DE VARIAÇÃO REVENDA PREÇO MÉDIO DISTRIBUIÇÃO  \
0                0.463                     0.012                    0.825   
1                0.399                     0.098                    0.763   
2                0.419                     0.070                     0.97   

  DESVIO PADRÃO DISTRIBUIÇÃO PREÇO MÍNIMO DISTRIBUIÇÃO  \
0                       0.11                    0.4201   
1                      0.088                    0.5013   
2                      0.095                    0.5614   

  PREÇO MÁXIMO DISTRIBUIÇÃO COEF DE VARIAÇÃO DISTRIBUIÇÃO  
0                    0.9666                         0.133  
1                      1.05                         0.115  
2                     1.161                         0.098

In [186]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 120823 entries, 0 to 120822
Data columns (total 18 columns):
 #   Column                         Non-Null Count   Dtype  
---  ------                         --------------   -----  
 0   DATA INICIAL                   120823 non-null  object 
 1   DATA FINAL                     120823 non-null  object 
 2   REGIÃO                         120823 non-null  object 
 3   ESTADO                         120823 non-null  object 
 4   PRODUTO                        120823 non-null  object 
 5   NÚMERO DE POSTOS PESQUISADOS   120823 non-null  int64  
 6   UNIDADE DE MEDIDA              120823 non-null  object 
 7   PREÇO MÉDIO REVENDA            120823 non-null  float64
 8   DESVIO PADRÃO REVENDA          120823 non-null  float64
 9   PREÇO MÍNIMO REVENDA           120823 non-null  float64
 10  PREÇO MÁXIMO REVENDA           120823 non-null  float64
 11  MARGEM MÉDIA REVENDA           120823 non-null  object 
 12  COEF DE VARIAÇÃO REVENDA      

In [187]:
#Conversão realizada coluna por coluna por ter valores diferentes
df['MARGEM MÉDIA REVENDA'] = df['MARGEM MÉDIA REVENDA'].apply(lambda x: float(x))
df['COEF DE VARIAÇÃO REVENDA'] = df['COEF DE VARIAÇÃO REVENDA'].apply(lambda x: float(x))
df['PREÇO MÉDIO REVENDA'] = df['PREÇO MÉDIO REVENDA'].apply(lambda x: float(x))
df['DESVIO PADRÃO REVENDA'] = df['DESVIO PADRÃO REVENDA'].apply(lambda x: float(x))
df['PREÇO MÍNIMO REVENDA'] = df['PREÇO MÍNIMO REVENDA'].apply(lambda x: float(x))
df['PREÇO MÁXIMO REVENDA'] = df['PREÇO MÁXIMO REVENDA'].apply(lambda x: float(x))
df['MARGEM MÉDIA REVENDA'] = df['MARGEM MÉDIA REVENDA'].apply(lambda x: float(x))
df['PREÇO MÉDIO DISTRIBUIÇÃO'] = df['PREÇO MÉDIO DISTRIBUIÇÃO'].apply(lambda x: float(x))
df['DESVIO PADRÃO DISTRIBUIÇÃO'] = df['DESVIO PADRÃO DISTRIBUIÇÃO'].apply(lambda x: float(x))
df['PREÇO MÍNIMO DISTRIBUIÇÃO'] = df['PREÇO MÍNIMO DISTRIBUIÇÃO'].apply(lambda x: float(x))
df['PREÇO MÁXIMO DISTRIBUIÇÃO'] = df['PREÇO MÁXIMO DISTRIBUIÇÃO'].apply(lambda x: float(x))
df['COEF DE VARIAÇÃO DISTRIBUIÇÃO'] = df['COEF DE VARIAÇÃO DISTRIBUIÇÃO'].apply(lambda x: float(x))

__Explorando conjunto de dados__

Realizando algumas perguntas para entender melhor o conjunto de dados

1: Qual o preço médio de revenda de gasolina das regiões do brasil? e se for considerado o período observado ano a ano?

In [188]:
#pergunta 1
df1 = df
df1.set_index('DATA INICIAL')
df1.groupby(['REGIÃO']).agg({'PREÇO MÉDIO REVENDA':sum}).sort_values('PREÇO MÉDIO REVENDA', ascending=False)

PREÇO MÉDIO REVENDA
REGIÃO                           
NORDESTE               444352.747
NORTE                  375379.522
CENTRO OESTE           217078.159
SUDESTE                196886.211
SUL                    152558.844

In [189]:
#pergunta 2
df2 = df
df2['Ano'] = df2['DATA INICIAL'].apply(lambda x: x.split('-')[0])
df2 = df2[['Ano','REGIÃO','PREÇO MÉDIO REVENDA']]
df2.head(3)
#dados = df2.groupby(['Ano','REGIÃO']).agg({'PREÇO MÉDIO REVENDA':sum})

Ano        REGIÃO  PREÇO MÉDIO REVENDA
0  2004  CENTRO OESTE                1.288
1  2004  CENTRO OESTE                1.162
2  2004  CENTRO OESTE                1.389

In [190]:
from sklearn.preprocessing import OneHotEncoder

In [191]:
ohe = OneHotEncoder()

In [192]:
variavel_ohe = df['REGIÃO'].values.reshape(-1,1)

ohe.fit(variavel_ohe)

OneHotEncoder()

In [193]:
ohe.transform(variavel_ohe)

<120823x5 sparse matrix of type '<class 'numpy.float64'>'
	with 120823 stored elements in Compressed Sparse Row format>

In [194]:
ohe.transform(variavel_ohe).toarray()

array([[1., 0., 0., 0., 0.],
       [1., 0., 0., 0., 0.],
       [1., 0., 0., 0., 0.],
       ...,
       [0., 0., 0., 1., 0.],
       [0., 1., 0., 0., 0.],
       [0., 0., 1., 0., 0.]])

In [195]:
ohe.categories_

[array(['CENTRO OESTE', 'NORDESTE', 'NORTE', 'SUDESTE', 'SUL'],
       dtype=object)]

In [196]:
dataframe_OHE = pd.DataFrame(ohe.transform(variavel_ohe).toarray(), 
                             columns = ohe.categories_[0].tolist())
print(dataframe_OHE.shape)
print(df2.shape)

(120823, 5)
(120823, 3)


In [197]:
dados = pd.concat([df2, dataframe_OHE], axis = 1)
dados.columns

Index(['Ano', 'REGIÃO', 'PREÇO MÉDIO REVENDA', 'CENTRO OESTE', 'NORDESTE',
       'NORTE', 'SUDESTE', 'SUL'],
      dtype='object')

In [209]:
dados = dados.groupby('Ano').sum()
dados['TOTAL'] = dados['CENTRO OESTE']+ dados['NORDESTE']+ dados['NORTE']+ dados['SUDESTE']+ dados['SUL']
dados['TOTAL_PREÇO MÉDIO REVENDA'] = dados['PREÇO MÉDIO REVENDA']/dados['TOTAL']

In [210]:
dados
#colunas de regiões é a quantidade de vendas

PREÇO MÉDIO REVENDA  CENTRO OESTE  NORDESTE   NORTE  SUDESTE    SUL  \
Ano                                                                         
2004            35117.510         624.0    1397.0  1092.0    624.0  468.0   
2005            51727.855         912.0    2070.0  1602.0    915.0  687.0   
2006            58835.338         972.0    2206.0  1696.0    986.0  738.0   
2007            58451.038         841.0    2367.0  1334.0   1105.0  828.0   
2008            59158.542         898.0    2278.0  1375.0   1088.0  814.0   
2009            59576.728         853.0    2245.0  1243.0   1084.0  807.0   
2010            65165.846         892.0    2405.0  1064.0   1215.0  913.0   
2011            66480.188         963.0    2164.0  1680.0    960.0  720.0   
2012            69281.644         982.0    2215.0  1715.0    984.0  738.0   
2013            75381.487        1184.0    2704.0  1849.0   1248.0  936.0   
2014            80018.459        1192.0    2704.0  1852.0   1248.0  936.0   
2015            88122.569        1164.0    2650.0  1817.0   1223.0  917.0   
2016           100123.682        1156.0    2706.0  1828.0   1247.0  936.0   
2017           109956.346        1154.0    2754.0  1856.0   1272.0  954.0   
2018           123496.839        1112.0    2702.0  1826.0   1248.0  935.0   
2019           125588.376        1104.0    2707.0  1819.0   1248.0  935.0   
2020           109379.690        1006.0    2368.0  1624.0   1099.0  825.0   
2021            50393.346         440.0    1029.0   717.0    476.0  357.0   

       TOTAL  TOTAL_PREÇO MÉDIO REVENDA  
Ano                                      
2004  4205.0                   8.351370  
2005  6186.0                   8.362085  
2006  6598.0                   8.917147  
2007  6475.0                   9.027187  
2008  6453.0                   9.167603  
2009  6232.0                   9.559809  
2010  6489.0                  10.042510  
2011  6487.0                  10.248218  
2012  6634.0                  10.443419  
2013  7921.0                   9.516663  
2014  7932.0                  10.088056  
2015  7771.0                  11.339927  
2016  7873.0                  12.717348  
2017  7990.0                  13.761745  
2018  7823.0                  15.786378  
2019  7813.0                  16.074283  
2020  6922.0                  15.801747  
2021  3019.0                  16.692066

# Essa parte dos dados estão inconsistentes